app/src/main/java/eu/faircode/email/DB.java

```diff
@@ -1034,25 +1050,39 @@ public static String encodeAddresses(Address[] addresses) {
             if (addresses == null)
                 return null;
             JSONArray jaddresses = new JSONArray();
-            for (Address address : addresses)
-                try {
-                    if (address instanceof InternetAddress) {
-                        String a = ((InternetAddress) address).getAddress();
-                        String p = ((InternetAddress) address).getPersonal();
-                        JSONObject jaddress = new JSONObject();
-                        if (a != null)
-                            jaddress.put("address", a);
-                        if (p != null)
-                            jaddress.put("personal", p);
-                        jaddresses.put(jaddress);
-                    } else {
-                        JSONObject jaddress = new JSONObject();
-                        jaddress.put("address", address.toString());
-                        jaddresses.put(jaddress);
-                    }
-                } catch (JSONException ex) {
-                    Log.e(ex);
+            for (Address address : addresses) {
+                JSONObject jaddress = inverseAddressCache.get(address);
+                if (jaddress == null) {
+                    try {
+                        jaddress = new JSONObject();
+                        InternetAddress internetAddress = null;
+                        if (address instanceof InternetAddress) {
+                            internetAddress = (InternetAddress) address;
+                            String a = internetAddress.getAddress();
+                            String p = internetAddress.getPersonal();
+                            if (a != null)
+                                jaddress.put("address", a);
+                            if (p != null)
+                                jaddress.put("personal", p);
+                        } else {
+                            jaddress.put("address", address.toString());
+                        }
+                        JSONAddress key = new JSONAddress(jaddress);
+                        if (!(internetAddress instanceof InternetAddressImpl)) {
+                            address = new InternetAddressImpl(internetAddress);
+                        }
+                        synchronized (addressCache) {
+                            addressCache.put(key, address);
+                            inverseAddressCache.put(address, key);
+                        }
+                    } catch (JSONException | UnsupportedEncodingException ex) {
+                        Log.e(ex);
+                    }
+                } else {
+                    jaddress = ((JSONAddress) jaddress).jsonObject;
                 }
+                jaddresses.put(jaddress);
+            }
```

Comment:
This makes the code more complex, but I don't think much is gained.

Reply:
Before optimizing `WorkerCleanup` it gained a lot, after that probably much less.

Reply:
The cleanup is not executed frequently and not user facing.